PYME/Acquire/Hardware/pco/pco_sdk_cam.py

```diff
@@ -0,0 +1,405 @@
+
+@author: zacsimile
+"""
+
+from PYME.Acquire.Hardware.Camera import Camera
+from PYME.Acquire.Hardware.pco import pco_sdk
+from PYME.Acquire import eventLog
+
+import numpy as np
+import logging
+logger = logging.getLogger(__name__)
+
+import ctypes
+import ctypes.wintypes
+
+k32_dll = ctypes.windll.kernel32  # lets us use the recommended WaitForSingleObject call (see pco.sdk)
+                                  # instead of the not-recommended-for-polling pco_sdk.get_buffer_status()
+
+timebase = {pco_sdk.PCO_TIMEBASE_NS : 1e-9, 
+            pco_sdk.PCO_TIMEBASE_US : 1e-6, 
+            pco_sdk.PCO_TIMEBASE_MS : 1e-3}  # Conversions 
+
+class camReg(object):
+    """
+    Keep track of the number of cameras initialised so we can initialise and
+    finalise the library.
+    """
+    numCameras = -1
+
+    @classmethod
+    def regCamera(cls):
+        if cls.numCameras == -1:
+            pco_sdk.reset_lib()
+
+        cls.numCameras += 1
+
+    @classmethod
+    def unregCamera(cls):
+        cls.numCameras -= 1
+        if cls.numCameras == 0:
+            # There appears to be no pco.sdk uninitialization
+            pass
+
+camReg.regCamera()  # initialize/reset the sdk
+
+MAX_BUFFERS = 16
+MAX_TIMEOUTS = 10
+
+class PcoSdkCam(Camera):
+    def __init__(self, camNum, debuglevel='off'):
+        Camera.__init__(self)
+        self._initalized = False
+        self.noiseProps = None
+
+    def Init(self):
+        self._handle = pco_sdk.open_camera()
+        camReg.regCamera()
+        self._desc = pco_sdk.get_camera_description(self._handle)
+
+        if pco_sdk.get_recording_state(self._handle) == pco_sdk.PCO_CAMERA_RUNNING:
+            pco_sdk.set_recording_state(self._handle, pco_sdk.PCO_CAMERA_STOPPED)
+        pco_sdk.reset_settings_to_default(self._handle)
+        _, err, _ = pco_sdk.get_camera_health_status(self._handle)
+        if err != 0:
+            self.Shutdown()
+            raise pco_sdk.PcoSdkException(f"Camera shutdown with error status {err}.")
+
+        self._bits_per_pixel = 16
+        self._integ_time = 0
+        self._delay_time = 0
+        self._electr_temp = 0
+        self._ccd_temp = 0
+        self.__default_n_buffers = 16  # max of 16
+        self._n_buffers = 0
+        self._buf_event = None
+        self._buf_addr = None
+        self._curr_buf = 0
+        self._initalized = True
+        self._recording = False
+        self._roi = None
+        self._timeout = 1000  # ms
+        self._n_buffered = 0
+        self._binning_x = 0
+        self._binning_y = 0
+        self._n_timeouts = 0
+        self.SetROI(1, 1, self.GetCCDWidth(), self.GetCCDHeight())
+        self.SetIntegTime(0.025)
+        self.SetAcquisitionMode(self.MODE_CONTINUOUS)
+        self._cam_type = pco_sdk.get_camera_type(self._handle)
+        self.SetHotPixelCorrectionMode(pco_sdk.PCO_HOTPIXELCORRECTION_OFF)
+
+    @property
+    def noise_properties(self):
+        return self.noiseProps
+
+    def ExpReady(self):
```

Comment:
As written, this blocks waiting for frames. When `ExpReady()` was polled from the GUI thread that would have been a major problem. If we can get away with it now , it's definitely cleaner.

Reply:
The general assumption made in the code however, is that `ExpReady()` is lightweight and non-blocking, so I'm worried that this will come back to bite us.

Reply:
I think you are right. It does work at the moment, but I suspect it can cause problems going faster. I could try the Zyla architecture. Maybe it will just work. The pco.sdk manual makes a big thing out of not using threading but it might be fine.

Reply:
I'd be tempted to just put a comment in the code, push it until it breaks, and then fix it when it becomes a problem.

Reply:
That would also be fine. I was able to take z-stacks with this today, so it's at least working well enough to do what I want for now.

Reply:
I've added a comment.